### ISSUE TITLE
envoy mobile should pass through connection: close header

### ISSUE THREAD

**Issue by alyssawilk** [envoyproxy/envoy-mobile#1540]
possibly other hop by hop headers.  this one I can take :-)

cc @carloseltuerto

**Comment by alyssawilk** [comment]
code is here: https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L439

offhand I think we want a knob "preserve hop by hop headers" which undoes 

  response_headers.removeConnection();
  response_headers.removeUpgrade();
  response_headers.removeTransferEncoding();
  response_headers.removeKeepAlive();
  response_headers.removeProxyConnection();

cc @DavidSchinazi @RyanTheOptimist in case naming if off (maybe one isn't hop by hop)

**Comment by alyssawilk** [comment]
hey @envoyproxy/api-shepherds  I'll send a WIP PR out for this shortly but it's kind of a weird case.
I believe the way Envoy mobile works, we need an HCM config knob to preserve the hop by hop headers.  It basically should never be used by Envoy-the-server and there's AFIK no validation checks we can put in place.
How do we feel about a field saying "don't use this for Envoy as a proxy"? Normally I'd think about putting it in some extensible any MobileConfig field but the code needs to go in the upstream HCM so that feels weird too.

**Comment by markdroth** [comment]
I think this is okay.  We can catch the problem at runtime when the client reads the resource from the xDS server, in which case it can NACK if it's set wrong.

**Comment by htuch** [comment]
Do we actually need configuration here? If the Envoy Mobile library has some way of conveying to HCM that this is in fact a linked-in client library (and hence there is no additional hop for removing headers), then we can avoid having to expose this as configuration.

**Comment by alyssawilk** [comment]
AFIK Envoy mobile uses standard factory creation.  @junr03 or @goaway could comment if there's an easy way to access the HCM directly but there's not that I know of.

**Comment by mattklein123** [comment]
> AFIK Envoy mobile uses standard factory creation. @junr03 or @goaway could comment if there's an easy way to access the HCM directly but there's not that I know of.

One thing we might consider is sub-classing HCM for Envoy Mobile and then instantiating that version? That would allow us to customize without adding a bunch of config that other people probably don't need.

**Comment by goaway** [comment]
I haven't thought deeply about it yet, but my gut reaction is I'm a little hesitant to subclass the HCM since it seems like it might increase the surface area for surprises when upstream changes (e.g., we implement some overriding business logic, and then the part we're overriding upstream changes to do something different, maybe critical, or moves parts around such that they're no longer overridden).

**Comment by goaway** [comment]
OTOH if all we're doing is setting up construction differently, and all the business logic remains in upstream Envoy, maybe that's not really as much of a can of worms.

**Comment by mattklein123** [comment]
> OTOH if all we're doing is setting up construction differently, and all the business logic remains in upstream Envoy, maybe that's not really as much of a can of worms.

Yes, exactly. I don't think it's that big of a deal as long as you have upstream tests for whatever you are setting/doing.

**Comment by mattklein123** [comment]
(You could even have the sub-class upstream with tests.)
